### DISCUSSION TITLE
Impact of Puma's idle_timeout on User-Created Threads in the Application

### DISCUSSION THREAD

**Discussion by fireva** [puma/puma#3274] [Q&A] [answered]
I have attempted to use the new feature included in Puma 6.4.0 (#3209 ) in my Rails application. For certain requests that take a considerable amount of time, I create a background process using the following code snippet:
`     Thread.new do`
`         # Some process that takes a long time `
`     end`
By doing this approach, I reduced the overall request time. The problem is, if no new requests are received within the specified time frame (idle_timeout config), Puma terminates this background process, even though it still running. Therefore, my questions are as follows:
1. Which worker is the 'Thread.new' process assigned to? Is it always to Worker 0?
2. Is it possible to maintain the worker, even if it hasn't received any new requests beyond the idle_timeout settings, while it is still processing certain background tasks?

I hope there is a solution for this, since it will help in reducing memory problem. Currently, after processing requests that consume a significant amount of RAM, the worker not returning the memory back to the OS. Any assistance is highly appreciated. Thank you.

Extra info:
OS : Ubuntu 20.04.6 LTS
Rails version: 4.2.11.3
Puma version : 6.4.0

**Comment by dentarg** [comment, marked as answer]
It is whatever worker are processing the request.

If you spawn background threads, I don't think it is a good idea to also use `idle_timeout` as you have discovered.

I think you should consider a background job system like https://github.com/resque/resque or https://github.com/sidekiq/sidekiq over just doing `Thread.new` for things that take a long time.

**Reply by fireva** [reply]
Okay, I guess I will check the other methods for background job then. Thanks!
